subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -962,4 +962,100 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec {
         """
     }
 
+    def "implementation of nested closure in decorated bean is tracked"() {
+        taskWithNestedDecoratedBean()
+        buildFile << """
+            extensions.create("bean", NestedBean.class)
+            
+            bean {
+                withAction { it.text = "hello" }
+            }
+            
+            task myTask(type: TaskWithNestedDecoratedAction) {
+                bean = project.bean
+            }
+        """
+
+        buildFile.makeOlder()
+
+        when:
+        run 'myTask'
+
+        then:
+        executedAndNotSkipped(':myTask')
+        file('build/tmp/myTask/output.txt').text == "hello"
+
+        when:
+        buildFile.text = buildFile.text.replace('it.text = "hello"', 'it.text = "changed"')
+        run 'myTask', '--info'
+
+        then:
+        executedAndNotSkipped(':myTask')
+        file('build/tmp/myTask/output.txt').text == "changed"
+        output.contains "Value of input property 'bean.action.class' has changed for task ':myTask'"
+    }
+
+    private TestFile nestedDecoratedBeanWithAction() {
```

Comment:
Why is this bean decorated? Isn't this just a `BeanWithNestedAction`? Maybe we should change the class name in the source (`NestedBean` -> `BeanWithNestedAction`), too.

Reply:
Well, it's decorated by the fact that it gets instantiated as an extension object, but that really has nothing to do with the bean itself, so good point - this isn't a well named method (or object for that matter).  I've renamed things a little bit to be more clear (see https://github.com/gradle/gradle/commit/eba30a33ac831caff6d5f3c307ead96a1bbca338).